esp-hal/src/dma/mod.rs

```diff
@@ -732,59 +730,66 @@ where
     }
 
     fn available(&mut self) -> usize {
-        if self.last_seen_handled_descriptor_ptr.is_null() {
-            self.last_seen_handled_descriptor_ptr = self.descriptors.as_mut_ptr();
-            return 0;
-        }
-
-        if self.available != 0 {
-            return self.available;
-        }
-
-        let descr_address = self.last_seen_handled_descriptor_ptr;
-        let mut dw0 = unsafe { descr_address.read_volatile() };
+        let last_in_dscr_ptr = self.rx_impl.last_in_dscr_address() as *mut DmaDescriptor;
+        let last_in_descr = unsafe { last_in_dscr_ptr.read_volatile() };
 
-        if dw0.owner() == Owner::Cpu && !dw0.is_empty() {
-            let descriptor_buffer = dw0.buffer;
-            let next_descriptor = dw0.next;
-
-            self.read_buffer_start = descriptor_buffer;
-            self.available = dw0.len();
-
-            dw0.set_owner(Owner::Dma);
-            dw0.set_length(0);
-            dw0.set_suc_eof(false);
+        if !last_in_descr.is_empty() && last_in_descr.owner() == Owner::Cpu {
+            let len = last_in_descr.len();
 
             unsafe {
-                descr_address.write_volatile(dw0);
+                last_in_dscr_ptr.write_volatile(last_in_descr);
             }
 
-            self.last_seen_handled_descriptor_ptr = if next_descriptor.is_null() {
-                self.descriptors.as_mut_ptr()
-            } else {
-                next_descriptor
-            };
+            if self.last_seen_handled_descriptor_ptr.is_null() {
+                self.last_seen_handled_descriptor_ptr = last_in_dscr_ptr;
+                self.read_buffer_start = last_in_descr.buffer;
+                self.available = len;
+                return self.available;
+            }
+
+            if self.last_seen_handled_descriptor_ptr != last_in_dscr_ptr {
+                self.available += len;
+                return self.available;
+            }
         }
 
         self.available
     }
 
     fn pop(&mut self, data: &mut [u8]) -> Result<usize, DmaError> {
```

Comment:
So just to confirm the behaviour of this method.

You're trying to read as many **whole** descriptors as you can into the provided `data` buffer?
Is the intent here to **not** support partial descriptor reads?

Reply:
Yes we only care about completely done descriptor buffers

Reply:
Supporting partial pop would be more complex because the descriptor buffers are not necessarily completely used - but I guess it's not much of a problem for users of the API. I added some documentation about these assumptions

Reply:
Ah no, I didn't mean reading a descriptor before the hardware is done writing to it. I meant reading part of of a fully written descriptor. At the moment descriptor sizes are about 4092, but a user maybe only want to pop 64 bytes at a time from the finished 4092 byte descriptor. (The tx channel has something like this I think)

Reply:
For TX it's somewhat easier to do - "problem" really is that DMA doesn't always fully fill the descriptor's buffer. We could certainly support it but not sure if it will ever be a real issue for user code

Reply:
Yeah for most drivers it won't be important. For cameras though. partial reading will be important for JPEGs since the application won't know the size in advance.
I'm in favour of leaving this feature for a future PR though. No existing drivers need it right now.
I'll keep this in mind for my next comments.